File: dynamic-datatable-board/src/dto/hrms/leaveRequest.ts
// src/dto/hrms/leaveRequest.ts

/** Full read model returned from API */
export interface LeaveRequestDto {
  id: number;
  employeeId: number;
  employeeName: string;
  leaveTypeId: number;
  leaveTypeName: string;
  startDate: string;       // ISO 8601 date string
  endDate: string;
  totalHours: number;
  reason: string;
  status: string;          // "Pending" | "Approved" | "Rejected"

  firstApproverId?: number;
  firstApproverName?: string;
  firstApprovalStatus: string;
  firstApprovalReason?: string;

  secondApproverId?: number;
  secondApproverName?: string;
  secondApprovalStatus: string;
  secondApprovalReason?: string;

  submittedByEmployeeId: number;
  submittedByEmployeeName?: string;
  onBehalfReason?: string;

  createdAt: string;
}

/** Payload for creating a new leave request */
export interface LeaveRequestFormDto {
  employeeId: number;
  leaveTypeId: number;
  startDate: string;
  endDate: string;
  reason: string;
  /** ID of the logged-in employee submitting on behalf */
  submittedByEmployeeId: number;
  onBehalfReason?: string;
}

/** Query params for paginated leave requests list */
export interface LeaveRequestQueryDto {
  pageIndex: number;
  pageSize: number;
  employeeId?: number;
  status?: string;
  leaveTypeId?: number;
  search?: string;
}

/** Request body for the simulation endpoint */
export interface LeaveSimulateRequestDto {
  startDate: string;
  endDate: string;
}

/** Response from the simulation endpoint */
export interface LeaveSimulateResponseDto {
  workingDays: number;
  totalHours: number;
  excludedDays: number;
  totalCalendarDays: number;
}

/** Payload for first / second level approval actions */
export interface LeaveApprovalDto {
  status: 'Approved' | 'Rejected';
  remarks?: string;
}

File: backend/Domain/Entities/HRMS/LeaveRequest.cs
namespace Domain.Entities.HRMS;

using System.ComponentModel.DataAnnotations.Schema;
using Domain.Common;

public class LeaveRequest : BaseEntity
{
    // The employee who is taking the leave
    public int EmployeeId { get; set; }
    public Employee? Employee { get; set; }

    public int LeaveTypeId { get; set; }
    public LeaveType? LeaveType { get; set; }

    public DateTime StartDate { get; set; }
    public DateTime EndDate { get; set; }

    // Calculated by simulation (excludes weekends, deducts lunch)
    public decimal TotalHours { get; set; }

    public string Reason { get; set; } = string.Empty;

    // Overall request status: Pending | Approved | Rejected
    public string Status { get; set; } = "Pending";

    // First-level approval
    public int? FirstApproverId { get; set; }
    [ForeignKey(nameof(FirstApproverId))]
    public Employee? FirstApprover { get; set; }
    public string FirstApprovalStatus { get; set; } = "Pending";

    // Second-level approval
    public int? SecondApproverId { get; set; }
    [ForeignKey(nameof(SecondApproverId))]
    public Employee? SecondApprover { get; set; }
    public string SecondApprovalStatus { get; set; } = "Pending";

    // Approval reasons
    public string? FirstApprovalReason { get; set; }
    public string? SecondApprovalReason { get; set; }

    // "Apply on Behalf" — the Employee who submitted this request
    // (may differ from EmployeeId when a manager applies on behalf of a subordinate)
    public int SubmittedByEmployeeId { get; set; }
    public string? OnBehalfReason { get; set; }
}
